spacerace-client/src/main.rs

```diff
@@ -0,0 +1,302 @@
+    /// Current orientation (rotation) of the ship.
+    orientation: f32,
+    /// Angular velocity (rotational speed) of the ship.
+    angular_velocity: f32,
+}
+
+/// A request to control the ship by applying thrust and rotation.
+#[derive(Debug, Serialize)]
+struct ControlRequest {
+    /// Password for authentication.
+    password: String,
+    /// Thrust command (1 = forward, -1 = backward, 0 = none).
+    thrust: i32,
+    /// Rotation command (positive = rotate right, negative = rotate left, 0 = none).
+    rotation: i32,
+}
+
+/// The server's response to a control request.
+#[derive(Debug, Deserialize)]
+struct ControlResponse {
+    /// Status of the control request. Typically "ok".
+    status: String,
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    // Set up logging with INFO level by default.
+    tracing_subscriber::fmt()
+        .with_max_level(tracing::Level::INFO)
+        .init();
+
+    // Read configuration from environment variables or use defaults.
+    let host = env::var("HOST").unwrap_or_else(|_| DEFAULT_HOST.to_string());
+    let player_name = env::var("PLAYER_NAME").unwrap_or_else(|_| "Player".to_string());
+    let player_password = env::var("PLAYER_PASSWORD").unwrap_or_else(|_| {
+        rand::distributions::Alphanumeric.sample_string(&mut rand::thread_rng(), 16)
+    });
+
+    let player_team = player_name.clone(); // Use player's name as team name.
+
+    let client = reqwest::Client::new();
+
+    // Register with the lobby
+    let lobby_req = LobbyRequest {
+        name: player_name.clone(),
+        team: player_team.clone(),
+        password: player_password.clone(),
+    };
+
+    let lobby_response: LobbyResponse = client
+        .post(format!("{}/lobby", host))
+        .json(&lobby_req)
+        .send()
+        .await?
+        .error_for_status()?
+        .json()
+        .await?;
+
+    info!("Registered with lobby: {:?}", lobby_response);
+
+    // Set up terminal for keyboard input
+    crossterm::terminal::enable_raw_mode()?;
+    let mut stdout = stdout();
+    execute!(stdout, cursor::Hide)?;
+
+    // Initial instructions
+    execute!(
+        stdout,
+        terminal::Clear(terminal::ClearType::All),
+        cursor::MoveTo(0, 0)
+    )?;
+    info!("Use arrow keys (or WASD) to control the ship. Press 'q' to quit.");
+    stdout.flush()?;
+
+    // Run the game loop
+    run_game_loop(client, player_password, host).await?;
+
+    Ok(())
+}
+
+/// Run the main game loop:
+/// - Periodically fetch the current game state from the server.
+/// - Display instructions.
+/// - Poll for keyboard input to send control commands.
+/// - Sleep briefly to avoid overwhelming the server and CPU.
+///
+/// # Arguments
+///
+/// * `client` - The HTTP client for sending requests.
+/// * `player_password` - The password of the current player.
+/// * `host` - The server host URL.
+async fn run_game_loop(
+    client: reqwest::Client,
+    player_password: String,
+    host: String,
```

Comment:
`host_url` or `server_url` would be a better name. Same for the environment variable.